src/katgpucbf/xbgpu/engine.py

```diff
@@ -431,7 +431,7 @@ async def gpu_proc_loop(self) -> None:  # noqa: D102
             await rx_item.async_wait_for_events()
 
             tx_item = await self._tx_free_item_queue.get()
-            await tx_item.async_wait_for_events()
+            tx_item.enqueue_wait_for_events(self._proc_command_queue)
```

Comment:
I don't believe this is safe. At line 458 you're modifying tx_item's GPU data from the host (via vkgdr), so the host needs to wait for the GPU to finish the previous work. In practice I think it happens to be safe because the sequencing in sender loop doesn't put the item onto the `_tx_free_item_queue` until it is complete, but no point in planting a land-mine.

Apart from correctness, I think it'll also cause a memory leak because `async_wait_for_events` is the point at which old events get cleaned out of the event list.